lib/benchmark/db/db.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package db
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+)
+
+// retrieveDatabaseCertificates issues user database certificates. Same flow as
+// `tsh db login`.
+func retrieveDatabaseCertificates(ctx context.Context, tc *client.TeleportClient, db types.Database, dbUser, dbName string) (tls.Certificate, error) {
+	profile, err := tc.ProfileStatus()
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	key, err := tc.IssueUserCertsWithMFA(ctx, client.ReissueParams{
+		RouteToCluster: tc.SiteName,
+		RouteToDatabase: proto.RouteToDatabase{
+			ServiceName: db.GetName(),
+			Protocol:    db.GetProtocol(),
+			Username:    dbUser,
+			Database:    dbName,
+		},
+		AccessRequests: profile.ActiveRequests.AccessRequests,
+	}, nil)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	rawCert, ok := key.DBTLSCerts[db.GetName()]
+	if !ok {
+		return tls.Certificate{}, trace.AccessDenied("failed to retrieve database certificates")
+	}
+
+	tlsCert, err := key.TLSCertificate(rawCert)
+	return tlsCert, trace.Wrap(err)
+}
+
+// getDatabase loads the database which the name matches.
+func getDatabase(ctx context.Context, tc *client.TeleportClient, serviceName string, protocol string) (types.Database, error) {
+	databases, err := tc.ListDatabases(ctx, &proto.ListResourcesRequest{
+		Namespace:           tc.Namespace,
+		ResourceType:        types.KindDatabaseServer,
+		PredicateExpression: fmt.Sprintf(`name == "%s" && resource.spec.protocol == "%s"`, serviceName, protocol),
```

Comment:
Recently @GavinFrazar have updated this logic in the context of resource collisions https://github.com/gravitational/teleport/pull/28505 

We should also support add support it to support `tsh bench postgres ...  postgres09` comand without providing full resource discovery  name: `tsh bench postgres  ... postgres09-rds-eu-central-1-278576220453`

Reply:
These bench commands are hidden right? If it's only for internal use, we might just skip adding that

Reply:
It might be better to enforce using the full resource name here, so we don't have any assumptions here (on name resolution). It could be a problem only if we don't return the full name anymore on `tsh db ls`.

Reply:
full name is available with `tsh db ls --verbose`. The intent for that UX change was to not annoy customers with long discovered resource names - as long as this is internal use I wouldn't worry about it